File: Hrp/cs/HrpTester.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Reflection;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading;

#if !UNITY_2018_4_OR_NEWER
[assembly: AssemblyTitle("HrpTester")]
[assembly: AssemblyProduct("HrpTester")]
[assembly: AssemblyCopyright("Copyright (C) 2019-2021 Advanced Media, Inc.")]
[assembly: ComVisible(false)]
[assembly: Guid("c7841911-a965-4734-9222-462422948eed")]
[assembly: AssemblyVersion("1.0.03")]
[assembly: AssemblyFileVersion("1.0.03")]
#endif

namespace HrpTester {

public class HrpTester : com.amivoice.hrp.HrpListener {
	#if UNITY_2018_4_OR_NEWER
	public static PseudoConsole Console;
	#endif

	public static void Main(string[] args) {
		// HTTP 音声認識サーバ URL
		string serverURL = null;
		// プロキシサーバ名
		string proxyServerName = null;
		// 音声データファイル名
		List<string> audioFileNames = new List<string>();
		// グラマファイル名
		string grammarFileNames = null;
		// プロファイル ID
		string profileId = null;
		// プロファイル登録単語
		string profileWords = null;
		// セグメンタプロパティ
		string segmenterProperties = null;
		// フィラー単語を保持するかどうか
		string keepFillerToken = null;
		// 認識中イベント発行間隔
		string resultUpdatedInterval = null;
		// 拡張情報
		string extension = null;
		// サービス認証キー文字列
		string authorization = null;
		// ドメイン ID (d=)
		string domainId = null;
		// 音声データ形式 (c=)
		string codec = null;
		// 認識結果タイプ (r=)
		string resultType = null;
		// 認識結果文字エンコーディング (e=)
		string resultEncoding = null;
		// サービス認証キー文字列 (u=)
		string serviceAuthorization = null;
		// 発話区間検出パラメータ文字列 (v=)
		string voiceDetection = null;
		// HTTP リクエストヘッダ群
		string acceptTopic = null;
		string contentType = null;
		string accept = null;
		// 処理タイプ ("s"／"sc"／"m"／"mc")
		string type = "mc";
		// 接続タイムアウト
		int connectTimeout = 5000;
		// 受信タイムアウト
		int receiveTimeout = 0;
		// 処理ループ (1～)
		int loop = 1;
		// スリープ時間
		int sleepTime = -2;
		// 詳細出力
		bool verbose = false;
		// 実装タイプ
		int implementation = 1;

		// 引数のチェック
		foreach (string arg in args) {
			if (arg.StartsWith("g=")) {
				grammarFileNames = arg.Substring(2);
			} else
			if (arg.StartsWith("i=")) {
				profileId = arg.Substring(2);
			} else
			if (arg.StartsWith("w=")) {
				profileWords = arg.Substring(2);
			} else
			if (arg.StartsWith("op=")) {
				segmenterProperties = arg.Substring(3);
			} else
			if (arg.StartsWith("of=")) {
				keepFillerToken = arg.Substring(3);
			} else
			if (arg.StartsWith("oi=")) {
				resultUpdatedInterval = arg.Substring(3);
			} else
			if (arg.StartsWith("oe=")) {
				extension = arg.Substring(3);
			} else
			if (arg.StartsWith("ou=")) {
				authorization = arg.Substring(3);
			} else
			if (arg.StartsWith("d=")) {
				domainId = arg.Substring(2);
			} else
			if (arg.StartsWith("c=")) {
				codec = arg.Substring(2);
			} else
			if (arg.StartsWith("r=")) {
				resultType = arg.Substring(2);
			} else
			if (arg.StartsWith("e=")) {
				resultEncoding = arg.Substring(2);
			} else
			if (arg.StartsWith("u=")) {
				serviceAuthorization = arg.Substring(2);
			} else
			if (arg.StartsWith("v=")) {
				voiceDetection = arg.Substring(2);
			} else
			if (arg.StartsWith("hd=")) {
				acceptTopic = arg.Substring(3);
			} else
			if (arg.StartsWith("hc=")) {
				contentType = arg.Substring(3);
			} else
			if (arg.StartsWith("hr=")) {
				accept = arg.Substring(3);
			} else
			if (arg.StartsWith("-t")) {
				type = arg.Substring(2);
			} else
			if (arg.StartsWith("-x")) {
				proxyServerName = arg.Substring(2);
			} else
			if (arg.StartsWith("-c")) {
				connectTimeout = int.Parse(arg.Substring(2));
			} else
			if (arg.StartsWith("-r")) {
				receiveTimeout = int.Parse(arg.Substring(2));
			} else
			if (arg.StartsWith("-l")) {
				loop = int.Parse(arg.Substring(2));
			} else
			if (arg.StartsWith("-e")) {
				if (arg.Length > 2) {
					sleepTime = int.Parse(arg.Substring(2));
				} else {
					sleepTime = -1;
				}
			} else
			if (arg.StartsWith("-v")) {
				verbose = true;
			} else
			if (arg.StartsWith("-2")) {
				implementation = 2;
			} else
			if (serverURL == null) {
				serverURL = arg;
			} else {
				audioFileNames.Add(arg);
			}
			if (verbose) {
				Console.WriteLine("DEBUG: " + arg);
			}
		}
		if (audioFileNames.Count == 0) {
			Console.WriteLine("Usage: HrpTester [<parameters/options>]");
			Console.WriteLine("                  <url>");
			Console.WriteLine("                   <audioFileName>...");
			Console.WriteLine("Parameters:");
			Console.WriteLine("  g=<grammarFileNames>");
			Console.WriteLine("  i=<profileId>");
			Console.WriteLine("  w=<profileWords>");
			Console.WriteLine("  op=<segmenterProperties>");
			Console.WriteLine("  of=<keepFillerToken>");
			Console.WriteLine("  oi=<resultUpdatedInterval>");
			Console.WriteLine("  oe=<extension>");
			Console.WriteLine("  ou=<authorization>");
			Console.WriteLine("  d=<domainId>");
			Console.WriteLine("  c=<codec>");
			Console.WriteLine("  r=<resultType>");
			Console.WriteLine("  e=<resultEncoding>");
			Console.WriteLine("  u=<serviceAuthorization>");
			Console.WriteLine("  v=<voiceDetection>");
			Console.WriteLine("  hd=<acceptTopic>            HTTP 'Accept-Topic' header instead d=<domainId>");
			Console.WriteLine("  hc=<contentType>            HTTP 'Content-Type' header instead c=<codec>");
			Console.WriteLine("  hr=<accept>                 HTTP 'Accept' header instead r=<resultType>");
			Console.WriteLine("Options:");
			Console.WriteLine("  -t<type>                    (default: -tmc)");
			Console.WriteLine("      <type>='s'              Single part & no chunked encoding");
			Console.WriteLine("      <type>='sc'             Single part & chunked encoding");
			Console.WriteLine("      <type>='m'              Multi parts & no chunked encoding");
			Console.WriteLine("      <type>='mc'             Multi parts & chunked encoding");
			Console.WriteLine("  -x<proxyServerName>         (default: -x)");
			Console.WriteLine("  -c<connectionTimeout>       (default: -c5000)");
			Console.WriteLine("  -r<receiveTimeout>          (default: -r0)");
			Console.WriteLine("  -l<loop>                    (default: -l1)");
			Console.WriteLine("  -e                          realtime simulation (default: -)");
			Console.WriteLine("  -v                          verbose output (default: -)");
			return;
		}

		// 開始時間の取得
		long startTime = DateTime.Now.Ticks * 100;

		// スリープ時間の計算
		if (sleepTime == -1) {
			if (codec != null) {
				if (codec.Equals("22K", StringComparison.OrdinalIgnoreCase) || codec.Equals("MSB22K", StringComparison.OrdinalIgnoreCase) || codec.Equals("LSB22K", StringComparison.OrdinalIgnoreCase)) {
					sleepTime = 4096 * 1000 / 2 / 22050;
				} else
				if (codec.Equals("16K", StringComparison.OrdinalIgnoreCase) || codec.Equals("MSB16K", StringComparison.OrdinalIgnoreCase) || codec.Equals("LSB16K", StringComparison.OrdinalIgnoreCase)) {
					sleepTime = 4096 * 1000 / 2 / 16000;
				} else
				if (codec.Equals("11K", StringComparison.OrdinalIgnoreCase) || codec.Equals("MSB11K", StringComparison.OrdinalIgnoreCase) || codec.Equals("LSB11K", StringComparison.OrdinalIgnoreCase)) {
					sleepTime = 4096 * 1000 / 2 / 11025;
				} else
				if (codec.Equals("8K", StringComparison.OrdinalIgnoreCase) || codec.Equals("MSB8K", StringComparison.OrdinalIgnoreCase) || codec.Equals("LSB8K", StringComparison.OrdinalIgnoreCase)) {
					sleepTime = 4096 * 1000 / 2 / 8000;
				} else
				if (codec.Equals("MULAW", StringComparison.OrdinalIgnoreCase) || codec.Equals("ALAW", StringComparison.OrdinalIgnoreCase)) {
					sleepTime = 4096 * 1000 / 1 / 8000;
				} else {
					sleepTime = 4096 * 1000 / 2 / 16000; // 16K が指定されたものとして計算...
				}
			} else {
				sleepTime = 4096 * 1000 / 2 / 16000; // 16K が指定されたものとして計算...
			}
		}

		// HTTP 音声認識サーバイベントリスナの作成
		com.amivoice.hrp.HrpListener listener = new HrpTester(verbose);

		// HTTP 音声認識サーバの初期化
		com.amivoice.hrp.Hrp hrp = com.amivoice.hrp.Hrp.construct(implementation);
		hrp.setListener(listener);
		hrp.setServerURL(serverURL);
		hrp.setProxyServerName(proxyServerName);
		hrp.setConnectTimeout(connectTimeout);
		hrp.setReceiveTimeout(receiveTimeout);
		hrp.setGrammarFileNames(grammarFileNames);
		hrp.setProfileId(profileId);
		hrp.setProfileWords(profileWords);
		hrp.setSegmenterProperties(segmenterProperties);
		hrp.setKeepFillerToken(keepFillerToken);
		hrp.setResultUpdatedInterval(resultUpdatedInterval);
		hrp.setExtension(extension);
		hrp.setAuthorization(authorization);
		hrp.setDomainId(domainId);
		hrp.setCodec(codec);
		hrp.setResultType(resultType);
		hrp.setResultEncoding(resultEncoding);
		hrp.setServiceAuthorization(serviceAuthorization);
		hrp.setVoiceDetection(voiceDetection);

		// HTTP 音声認識サーバへの接続
		if (!hrp.connect()) {
			if (!verbose) {
				Console.WriteLine(hrp.getLastMessage());
			}
			Console.WriteLine("HTTP 音声認識サーバ " + serverURL + " への接続に失敗しました。");
			return;
		}

		try {
			// <!--
			if (verbose) {
				Console.WriteLine();
				Console.WriteLine("======================= " + com.amivoice.hrp.Hrp.getVersion());
			}
			// -->
			for (int i = 0; i < loop; i++) {
				foreach (string audioFileName in audioFileNames) {
					// <!--
					if (verbose) {
						Console.WriteLine();
						if (loop > 1) {
							Console.WriteLine("----------------------- [" + (i + 1) + "] " + audioFileName);
						} else {
							Console.WriteLine("----------------------- " + audioFileName);
						}
						Console.WriteLine();
					}
					// -->

					// <!-- HTTP リクエストヘッダ群
					if (acceptTopic != null) {
						hrp.setAcceptTopic(acceptTopic);
					}
					if (contentType != null) {
						if (contentType.Equals("*")) {
							if (audioFileName.EndsWith(".a08")) {
								if (audioFileName.IndexOf("msb") != -1) {
									hrp.setContentType("audio/x-pcm;rate=8000;bit=16;byte-order=big-endian");
								} else {
									hrp.setContentType("audio/x-pcm;rate=8000;bit=16;byte-order=little-endian");
								}
							} else
							if (audioFileName.EndsWith(".a16") || audioFileName.EndsWith(".adc")) {
								if (audioFileName.IndexOf("msb") != -1) {
									hrp.setContentType("audio/x-pcm;rate=16000;bit=16;byte-order=big-endian");
								} else {
									hrp.setContentType("audio/x-pcm;rate=16000;bit=16;byte-order=little-endian");
								}
							} else
							if (audioFileName.EndsWith(".mu")) {
								hrp.setContentType("audio/x-mulaw;rate=8000");
							} else
							if (audioFileName.EndsWith(".alaw")) {
								hrp.setContentType("audio/x-alaw;rate=8000");
							}
						} else {
							hrp.setContentType(contentType);
						}
					}
					if (accept != null) {
						hrp.setAccept(accept);
					}
					// -->

					// 音声データファイルの総バイト数の取得
					long audioDataBytes = new FileInfo(audioFileName).Length;
					if (audioDataBytes == 0) {
						Console.WriteLine("音声データファイル " + audioFileName + " の読み込みに失敗しました。");
						break;
					}

					// HTTP 音声認識サーバへの音声データの送信の開始
					if (!hrp.feedDataResume(type, audioDataBytes)) {
						if (!verbose) {
							Console.WriteLine(hrp.getLastMessage());
						}
						Console.WriteLine("HTTP 音声認識サーバへの音声データの送信の開始に失敗しました。");
						break;
					}

					try {
						using (FileStream audioStream = new FileStream(audioFileName, FileMode.Open, FileAccess.Read)) {
							// 音声データファイルからの音声データの読み込み
							byte[] audioData = new byte[4096];
							int audioDataReadBytes = audioStream.Read(audioData, 0, audioData.Length);
							while (audioDataReadBytes > 0) {
								// スリープ時間が計算されているかどうかのチェック
								if (sleepTime >= 0) {
									// スリープ時間が計算されている場合...
									// 微小時間のスリープ
									hrp.sleep(sleepTime);
								} else {
									// スリープ時間が計算されていない場合...
									// 微小時間のスリープ
									hrp.sleep(1);
								}

								// HTTP 音声認識サーバへの音声データの送信
								if (!hrp.feedData(audioData, 0, audioDataReadBytes)) {
									if (!verbose) {
										Console.WriteLine(hrp.getLastMessage());
									}
									Console.WriteLine("HTTP 音声認識サーバへの音声データの送信に失敗しました。");
									break;
								}

								// 音声データファイルからの音声データの読み込み
								audioDataReadBytes = audioStream.Read(audioData, 0, audioData.Length);
							}
						}
					} catch (IOException) {
						Console.WriteLine("音声データファイル " + audioFileName + " の読み込みに失敗しました。");
					}

					// HTTP 音声認識サーバへの音声データの送信の完了
					if (!hrp.feedDataPause()) {
						if (!verbose) {
							Console.WriteLine(hrp.getLastMessage());
						}
						Console.WriteLine("HTTP 音声認識サーバへの音声データの送信の完了に失敗しました。");
						break;
					}
				}
			}
			// <!--
			if (verbose) {
				Console.WriteLine();
				Console.WriteLine("=======================");
				Console.WriteLine();
			}
			// -->
		} finally {
			// HTTP 音声認識サーバからの切断
			hrp.disconnect();
		}

		// 終了時間の取得
		long endTime = DateTime.Now.Ticks * 100;

		// 経過時間の表示
		Console.WriteLine("INFO: elapsed time: " + ((float)(endTime - startTime) / 1000000.0f) + " [msec.]");
	}

	private bool verbose_;

	public HrpTester(bool verbose) {
		verbose_ = verbose;
	}

	public void resultCreated(string sessionId) {
		Console.WriteLine("C " + sessionId);
	}

	public void resultUpdated(string result) {
		Console.WriteLine("U " + result);
		string text = text_(result);
		if (text != null) {
			Console.WriteLine("   -> " + text);
		}
	}

	public void resultFinalized(string result) {
		Console.WriteLine("F " + result);
		string text = text_(result);
		if (text != null) {
			Console.WriteLine("   -> " + text);
		}
	}

	public void TRACE(string message) {
		if (verbose_) {
			Console.WriteLine(message);
		}
	}

	private string text_(string result) {
		int index = result.LastIndexOf(",\"text\":\"");
		if (index == -1) {
			return null;
		}
		index += 9;
		int resultLength = result.Length;
		StringBuilder buffer = new StringBuilder();
		int c = (index >= resultLength) ? 0 : result[index++];
		while (c != 0) {
			if (c == '"') {
				break;
			}
			if (c == '\\') {
				c = (index >= resultLength) ? 0 : result[index++];
				if (c == 0) {
					return null;
				}
				if (c == '"' || c == '\\' || c == '/') {
					buffer.Append((char)c);
				} else
				if (c == 'b' || c == 'f' || c == 'n' || c == 'r' || c == 't') {
				} else
				if (c == 'u') {
					int c0 = (index >= resultLength) ? 0 : result[index++];
					int c1 = (index >= resultLength) ? 0 : result[index++];
					int c2 = (index >= resultLength) ? 0 : result[index++];
					int c3 = (index >= resultLength) ? 0 : result[index++];
					if (c0 >= '0' && c0 <= '9') {c0 -= '0';} else if (c0 >= 'A' && c0 <= 'F') {c0 -= 'A' - 10;} else if (c0 >= 'a' && c0 <= 'f') {c0 -= 'a' - 10;} else {c0 = -1;}
					if (c1 >= '0' && c1 <= '9') {c1 -= '0';} else if (c1 >= 'A' && c1 <= 'F') {c1 -= 'A' - 10;} else if (c1 >= 'a' && c1 <= 'f') {c1 -= 'a' - 10;} else {c1 = -1;}
					if (c2 >= '0' && c2 <= '9') {c2 -= '0';} else if (c2 >= 'A' && c2 <= 'F') {c2 -= 'A' - 10;} else if (c2 >= 'a' && c2 <= 'f') {c2 -= 'a' - 10;} else {c2 = -1;}
					if (c3 >= '0' && c3 <= '9') {c3 -= '0';} else if (c3 >= 'A' && c3 <= 'F') {c3 -= 'A' - 10;} else if (c3 >= 'a' && c3 <= 'f') {c3 -= 'a' - 10;} else {c3 = -1;}
					if (c0 == -1 || c1 == -1 || c2 == -1 || c3 == -1) {
						return null;
					}
					buffer.Append((char)((c0 << 12) | (c1 << 8) | (c2 << 4) | c3));
				} else {
					return null;
				}
			} else {
				buffer.Append((char)c);
			}
			c = (index >= resultLength) ? 0 : result[index++];
		}
		return buffer.ToString();
	}
}

} // namespace

File: Hrp/python/HrpTester.py
# encoding: UTF-8
import os
import sys
import time
import json
# <!-- バイトコードキャッシュファイルの作成を抑制するために...
sys.dont_write_bytecode = True
# -->
import com.amivoice.hrp.Hrp
import com.amivoice.hrp.HrpListener

class HrpTester(com.amivoice.hrp.HrpListener):
	@staticmethod
	def main(args):
		# HTTP 音声認識サーバ URL
		serverURL = None
		# プロキシサーバ名
		proxyServerName = None
		# 音声データファイル名
		audioFileNames = []
		# グラマファイル名
		grammarFileNames = None
		# プロファイル ID
		profileId = None
		# プロファイル登録単語
		profileWords = None
		# セグメンタプロパティ
		segmenterProperties = None
		# フィラー単語を保持するかどうか
		keepFillerToken = None
		# 認識中イベント発行間隔
		resultUpdatedInterval = None
		# 拡張情報
		extension = None
		# サービス認証キー文字列
		authorization = None
		# ドメイン ID (d=)
		domainId = None
		# 音声データ形式 (c=)
		codec = None
		# 認識結果タイプ (r=)
		resultType = None
		# 認識結果文字エンコーディング (e=)
		resultEncoding = None
		# サービス認証キー文字列 (u=)
		serviceAuthorization = None
		# 発話区間検出パラメータ文字列 (v=)
		voiceDetection = None
		# HTTP リクエストヘッダ群
		acceptTopic = None
		contentType = None
		accept = None
		# 処理タイプ ("s"／"sc"／"m"／"mc")
		type = "mc"
		# 接続タイムアウト
		connectTimeout = 5000
		# 受信タイムアウト
		receiveTimeout = 0
		# 処理ループ (1～)
		loop = 1
		# スリープ時間
		sleepTime = -2
		# 詳細出力
		verbose = False
		# 実装タイプ
		implementation = 1

		# 引数のチェック
		for arg in args:
			if arg.startswith("g="):
				grammarFileNames = arg[2:]
			elif arg.startswith("i="):
				profileId = arg[2:]
			elif arg.startswith("w="):
				profileWords = arg[2:]
			elif arg.startswith("op="):
				segmenterProperties = arg[3:]
			elif arg.startswith("of="):
				keepFillerToken = arg[3:]
			elif arg.startswith("oi="):
				resultUpdatedInterval = arg[3:]
			elif arg.startswith("oe="):
				extension = arg[3:]
			elif arg.startswith("ou="):
				authorization = arg[3:]
			elif arg.startswith("d="):
				domainId = arg[2:]
			elif arg.startswith("c="):
				codec = arg[2:]
			elif arg.startswith("r="):
				resultType = arg[2:]
			elif arg.startswith("e="):
				resultEncoding = arg[2:]
			elif arg.startswith("u="):
				serviceAuthorization = arg[2:]
			elif arg.startswith("v="):
				voiceDetection = arg[2:]
			elif arg.startswith("hd="):
				acceptTopic = arg[3:]
			elif arg.startswith("hc="):
				contentType = arg[3:]
			elif arg.startswith("hr="):
				accept = arg[3:]
			elif arg.startswith("-t"):
				type = arg[2:]
			elif arg.startswith("-x"):
				proxyServerName = arg[2:]
			elif arg.startswith("-c"):
				connectTimeout = int(arg[2:])
			elif arg.startswith("-r"):
				receiveTimeout = int(arg[2:])
			elif arg.startswith("-l"):
				loop = int(arg[2:])
			elif arg.startswith("-e"):
				if len(arg) > 2:
					sleepTime = int(arg[2:])
				else:
					sleepTime = -1
			elif arg.startswith("-v"):
				verbose = True
			elif arg.startswith("-2"):
				implementation = 2
			elif serverURL == None:
				serverURL = arg
			else:
				audioFileNames.append(arg)
			if verbose:
				print("DEBUG: %s" % arg)
		if len(audioFileNames) == 0:
			print("Usage: python HrpTester.py [<parameters/options>]")
			print("                            <url>")
			print("                             <audioFileName>...")
			print("Parameters:")
			print("  g=<grammarFileNames>")
			print("  i=<profileId>")
			print("  w=<profileWords>")
			print("  op=<segmenterProperties>")
			print("  of=<keepFillerToken>")
			print("  oi=<resultUpdatedInterval>")
			print("  oe=<extension>")
			print("  ou=<authorization>")
			print("  d=<domainId>")
			print("  c=<codec>")
			print("  r=<resultType>")
			print("  e=<resultEncoding>")
			print("  u=<serviceAuthorization>")
			print("  v=<voiceDetection>")
			print("  hd=<acceptTopic>            HTTP 'Accept-Topic' header instead d=<domainId>")
			print("  hc=<contentType>            HTTP 'Content-Type' header instead c=<codec>")
			print("  hr=<accept>                 HTTP 'Accept' header instead r=<resultType>")
			print("Options:")
			print("  -t<type>                    (default: -tmc)")
			print("      <type>='s'              Single part & no chunked encoding")
			print("      <type>='sc'             Single part & chunked encoding")
			print("      <type>='m'              Multi parts & no chunked encoding")
			print("      <type>='mc'             Multi parts & chunked encoding")
			print("  -x<proxyServerName>         (default: -x)")
			print("  -c<connectionTimeout>       (default: -c5000)")
			print("  -r<receiveTimeout>          (default: -r0)")
			print("  -l<loop>                    (default: -l1)")
			print("  -e                          realtime simulation (default: -)")
			print("  -v                          verbose output (default: -)")
			return

		# 開始時間の取得
		startTime = time.time()

		# スリープ時間の計算
		if sleepTime == -1:
			if codec != None:
				if codec.upper() == "22K" or codec.upper() == "MSB22K" or codec.upper() == "LSB22K":
					sleepTime = 4096 * 1000 / 2 / 22050
				elif codec.upper() == "16K" or codec.upper() == "MSB16K" or codec.upper() == "LSB16K":
					sleepTime = 4096 * 1000 / 2 / 16000
				elif codec.upper() == "11K" or codec.upper() == "MSB11K" or codec.upper() == "LSB11K":
					sleepTime = 4096 * 1000 / 2 / 11025
				elif codec.upper() == "8K" or codec.upper() == "MSB8K" or codec.upper() == "LSB8K":
					sleepTime = 4096 * 1000 / 2 / 8000
				elif codec.upper() == "MULAW" or codec.upper() == "ALAW":
					sleepTime = 4096 * 1000 / 1 / 8000
				else:
					sleepTime = 4096 * 1000 / 2 / 16000 # 16K が指定されたものとして計算...
			else:
				sleepTime = 4096 * 1000 / 2 / 16000 # 16K が指定されたものとして計算...

		# HTTP 音声認識サーバイベントリスナの作成
		listener = HrpTester(verbose)

		# HTTP 音声認識サーバの初期化
		hrp = com.amivoice.hrp.Hrp.construct(implementation)
		hrp.setListener(listener)
		hrp.setServerURL(serverURL)
		hrp.setProxyServerName(proxyServerName)
		hrp.setConnectTimeout(connectTimeout)
		hrp.setReceiveTimeout(receiveTimeout)
		hrp.setGrammarFileNames(grammarFileNames)
		hrp.setProfileId(profileId)
		hrp.setProfileWords(profileWords)
		hrp.setSegmenterProperties(segmenterProperties)
		hrp.setKeepFillerToken(keepFillerToken)
		hrp.setResultUpdatedInterval(resultUpdatedInterval)
		hrp.setExtension(extension)
		hrp.setAuthorization(authorization)
		hrp.setDomainId(domainId)
		hrp.setCodec(codec)
		hrp.setResultType(resultType)
		hrp.setResultEncoding(resultEncoding)
		hrp.setServiceAuthorization(serviceAuthorization)
		hrp.setVoiceDetection(voiceDetection)

		# HTTP 音声認識サーバへの接続
		if not hrp.connect():
			if not verbose:
				print(hrp.getLastMessage())
			print(u"HTTP 音声認識サーバ %s への接続に失敗しました。" % serverURL)
			return

		try:
			# <!--
			if verbose:
				print("")
				print("======================= " + com.amivoice.hrp.Hrp.getVersion())
			# -->
			for i in range(0, loop):
				for audioFileName in audioFileNames:
					# <!--
					if verbose:
						print("")
						if loop > 1:
							print("----------------------- [%d] %s" % (i + 1, audioFileName))
						else:
							print("----------------------- %s" % audioFileName)
						print("")
					# -->

					# <!-- HTTP リクエストヘッダ群
					if acceptTopic != None:
						hrp.setAcceptTopic(acceptTopic)
					if contentType != None:
						if contentType == "*":
							if audioFileName.endswith(".a08"):
								if audioFileName.find("msb") != -1:
									hrp.setContentType("audio/x-pcm;rate=8000;bit=16;byte-order=big-endian")
								else:
									hrp.setContentType("audio/x-pcm;rate=8000;bit=16;byte-order=little-endian")
							elif audioFileName.endswith(".a16") or audioFileName.endswith(".adc"):
								if audioFileName.find("msb") != -1:
									hrp.setContentType("audio/x-pcm;rate=16000;bit=16;byte-order=big-endian")
								else:
									hrp.setContentType("audio/x-pcm;rate=16000;bit=16;byte-order=little-endian")
							elif audioFileName.endswith(".mu"):
								hrp.setContentType("audio/x-mulaw;rate=8000")
							elif audioFileName.endswith(".alaw"):
								hrp.setContentType("audio/x-alaw;rate=8000")
						else:
							hrp.setContentType(contentType)
					if accept != None:
						hrp.setAccept(accept)
					# -->

					# 音声データファイルの総バイト数の取得
					try:
						audioDataBytes = os.path.getsize(audioFileName)
					except:
						print(u"音声データファイル %s の読み込みに失敗しました。" % audioFileName)
						break

					# HTTP 音声認識サーバへの音声データの送信の開始
					if not hrp.feedDataResume(type, audioDataBytes):
						if not verbose:
							print(hrp.getLastMessage())
						print(u"HTTP 音声認識サーバへの音声データの送信の開始に失敗しました。")
						break

					try:
						with open(audioFileName, "rb") as audioStream:
							# 音声データファイルからの音声データの読み込み
							audioData = audioStream.read(4096)
							while len(audioData) > 0:
								# スリープ時間が計算されているかどうかのチェック
								if sleepTime >= 0:
									# スリープ時間が計算されている場合...
									# 微小時間のスリープ
									hrp.sleep(sleepTime)
								else:
									# スリープ時間が計算されていない場合...
									# 微小時間のスリープ
									hrp.sleep(1)

								# HTTP 音声認識サーバへの音声データの送信
								if not hrp.feedData(audioData, 0, len(audioData)):
									if not verbose:
										print(hrp.getLastMessage())
									print(u"HTTP 音声認識サーバへの音声データの送信に失敗しました。")
									break

								# 音声データファイルからの音声データの読み込み
								audioData = audioStream.read(4096)
					except:
						print(u"音声データファイル %s の読み込みに失敗しました。" % audioFileName)

					# HTTP 音声認識サーバへの音声データの送信の完了
					if not hrp.feedDataPause():
						if not verbose:
							print(hrp.getLastMessage())
						print(u"HTTP 音声認識サーバへの音声データの送信の完了に失敗しました。")
						break
			# <!--
			if verbose:
				print("")
				print("=======================")
				print("")
			# -->
		finally:
			# HTTP 音声認識サーバからの切断
			hrp.disconnect()

		# 終了時間の取得
		endTime = time.time()

		# 経過時間の表示
		print("INFO: elapsed time: %.3f [msec.]" % ((endTime - startTime) * 1000.0))

	def __init__(self, verbose):
		self.verbose_ = verbose

	def resultCreated(self, sessionId):
		print("C %s" % sessionId)

	def resultUpdated(self, result):
		print("U %s" % result)
		text = self.text_(result)
		if text != None:
			print("   -> %s" % text)

	def resultFinalized(self, result):
		print("F %s" % result)
		text = self.text_(result)
		if text != None:
			print("   -> %s" % text)

	def TRACE(self, message):
		if self.verbose_:
			print(message)

	def text_(self, result):
		try:
			return json.loads(result)["text"]
		except:
			return None

HrpTester.main(sys.argv[1:])
